libsharedringbuffer/include/shared_ringbuffer/shared_ringbuffer.h

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#pragma once 
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sel4utils/sel4_zf_logif.h>
+#include <utils/util.h>
+#include <shared_ringbuffer/gen_config.h>
+
+/* Function pointer to be used to 'notify' components on either end of the shared memory */
+typedef void (*notify_fn)(void);
+
+/* Buffer descriptor */
+typedef struct buff_desc {
+    uintptr_t encoded_addr; /* encoded dma addresses */
+    unsigned int len; /* associated memory lengths */
+    void *cookie; /* index into client side metadata */
+} buff_desc_t;
+
+/* Circular buffer containing descriptors */
+typedef struct ring_buffer {
+    uint32_t write_idx;
```

Comment:
Should be `size_t` also and not fix is to 32-bit (even is 32-bit might be practically sufficient.

Reply:
I thought that size_t would be reserved for lengths/sizes of things. This is just an index into the array

Reply:
For portable code, my rule of thumb usually is, that it's either `unsigned int` (as C's native integer) or `size_t` (to support the maximum range, which might be more than `unsigned int` covers) then. If specific fix-size types are used, this should have a good reason or some explanation that we want to ensure something - but this would also become compile time asserts then.
When it comes to structures, a reason might be that the structure's memory layout is important because it get's serialized/deserialized and thus has to work across different architectures.